networkx/algorithms/approximation/fvs.py

```diff
@@ -0,0 +1,222 @@
+
+    Raises
+    ------
+    NetworkXNotImplemented
+        If `G` is directed or a multigraph.
+
+    Notes
+    -----
+    This algorithm is from [1]_. It returns a feedback vertex set whose
+    total weight is at most twice the weight of the true minimum
+    feedback vertex set. The worst-case running time of the algorithm is
+    :math:`O(n^2)`, where *n* is the number of nodes in the graph,
+    ignoring polylogarithmic factors.
+
+    References
+    ----------
+    .. [1] Bafna, Vineet, Piotr Berman, and Toshihiro Fujito.
+       "A 2-approximation algorithm for the undirected
+       feedback vertex set problem." *SIAM Journal on Discrete Mathematics*
+       12.3 (1999): 289-297. <https://dx.doi.org/10.1137/S0895480196305124>
+
+    """
+
+    def move_and_cleanup(G, F, weights, stack=None):
+        """Move all nodes with weight zero from G to F.
+
+        Each node of weight zero moved from `G` to `F` is recorded in
+        `stack`, if given. After removing the nodes of weight zero from
+        `G` and adding them to `F`, this function modifies `G` to be its
+        two-core. This also clears the nodes of weight zero from the
+        `weights` dictionary as well.
+
+        """
+        # Move all weight zero nodes from G to F, pushing each onto the
+        # stack if necessary.
+        weight_zero_nodes = {v for v in G if weights[v] == 0}
+        G.remove_nodes_from(weight_zero_nodes)
+        F |= weight_zero_nodes
+        if stack is not None:
+            stack.extend(weight_zero_nodes)
+        # Compute the two-core and remove each node not in the two-core
+        # from the dictionary of weights.
+        before = set(G)
+        G = nx.k_core(G, k=2)
+        for v in before - set(G):
+            del weights[v]
+
+    # Store the original graph, since G will be modified.
+    G_original, G = G, G.copy(with_data=False)
+    # Create a separate dictionary of weights, since the weights of each
+    # node will be updated in the main loop.
+    weights = Counter({u: G.node[u].get(weight, 1) for u in G})
+    # Initially, F is the empty set. As the algorithm progresses, nodes
+    # will be removed from the graph and added to F. At the end of the
+    # algorithm, F will contain a feedback vertex set.
+    F = set()
+    # The stack stores the order in which nodes are removed from the graph.
+    stack = []
+    # First, remove all nodes of weight zero from the graph and add them
+    # to F. Clean up the graph by removing any dangling tails.
+    move_and_cleanup(G, F, weights)
+    while G:
+        # Try to find a semidisjoint cycle in the graph.
+        try:
+            C = next(semidisjoint_cycles(G))
+        # If there are no semidisjoint cycles, update the degree-scaled
+        # weights of all nodes.
+        except StopIteration:
+            gamma = min(weights[v] / (d - 1) for v, d in G.degree())
+            weights -= {v: gamma * (d - 1) for v, d in G.degree()}
```

Comment:
Tried running this to see what the problem was - my problem (running Python 2.7 - don't know if this is different in other versions) was this line where we're trying to subtract a dictionary from a Counter. Was this dictionary meant to be a Counter as well?